lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,582 @@
+        Box::new(
+            (0..self.mappings.internal_to_external.len() as PointOffsetType)
+                .filter(move |i| !self.deleted[*i as usize]),
+        )
+    }
+
+    fn iter_from(
+        &self,
+        external_id: Option<PointIdType>,
+    ) -> Box<dyn Iterator<Item = (PointIdType, PointOffsetType)> + '_> {
+        let full_num_iter = || {
+            self.mappings
+                .external_to_internal_num
+                .iter()
+                .map(|(k, v)| (PointIdType::NumId(*k), *v))
+        };
+        let offset_num_iter = |offset: u64| {
+            self.mappings
+                .external_to_internal_num
+                .range(offset..)
+                .map(|(k, v)| (PointIdType::NumId(*k), *v))
+        };
+        let full_uuid_iter = || {
+            self.mappings
+                .external_to_internal_uuid
+                .iter()
+                .map(|(k, v)| (PointIdType::Uuid(*k), *v))
+        };
+        let offset_uuid_iter = |offset: Uuid| {
+            self.mappings
+                .external_to_internal_uuid
+                .range(offset..)
+                .map(|(k, v)| (PointIdType::Uuid(*k), *v))
+        };
+
+        match external_id {
+            None => {
+                let iter_num = full_num_iter();
+                let iter_uuid = full_uuid_iter();
+                // order is important here, we want to iterate over the u64 ids first
+                Box::new(iter_num.chain(iter_uuid))
+            }
+            Some(offset) => match offset {
+                PointIdType::NumId(idx) => {
+                    // Because u64 keys are less that uuid key, we can just use the full iterator for uuid
+                    let iter_num = offset_num_iter(idx);
+                    let iter_uuid = full_uuid_iter();
+                    // order is important here, we want to iterate over the u64 ids first
+                    Box::new(iter_num.chain(iter_uuid))
+                }
+                PointIdType::Uuid(uuid) => {
+                    // if offset is a uuid, we can only iterate over uuids
+                    Box::new(offset_uuid_iter(uuid))
+                }
+            },
+        }
+    }
+
+    fn iter_ids(&self) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        self.iter_internal()
+    }
+
+    /// Creates a flusher function, that writes the deleted points bitvec to disk.
+    fn mapping_flusher(&self) -> Flusher {
+        self.deleted_wrapper.flusher()
+    }
+
+    /// Creates a flusher function, that writes the points versions to disk.
+    fn versions_flusher(&self) -> Flusher {
+        self.internal_to_version_wrapper.flusher()
+    }
+
+    fn total_point_count(&self) -> usize {
+        self.mappings.internal_to_external.len()
+    }
+
+    fn available_point_count(&self) -> usize {
+        self.mappings.external_to_internal_num.len() + self.mappings.external_to_internal_uuid.len()
+    }
```

Comment:
I don't think this is correct. It should decrease if we mark a point is deleted.

Or does that indeed happen and am I missing an important step somewhere else?

Reply:
Looks correct to be, since new in-memory representation is equivalent to the SimpleIdTracker

Reply:
hm, maybe we should modify it then?

Reply:
Then we'd also need to flush and memory map them too. What do you think about counting the amount of deleted points and just subtract them? We could simply store this value as header in the memory mapped `.deleted` file.
I think this would be more efficient and cleaner.

Reply:
I would like that approach better.

We have another case where we do this. There we don't store the count, but count again on start because we may not be able to reliably store it: https://github.com/qdrant/qdrant/blob/8f1d84e40427f8bfd946edfac7a1b03656e37d41/lib/segment/src/vector_storage/dense/appendable_mmap_dense_vector_storage.rs#L77-L78

Reply:
Oh you're right. I'd still prefer the deleted-count tracker, unless of course you prefer deleting points from the BTreeMap.